Event Device Support (`evdev`)

votti opened the pull request:
## Background
As described in  #1942, a plugin to support to setup listeners to event devices such as USB controller, joysticks or keyboards would be useful.

## Implementation
This implementation builds on the `EvDevKeyListener` and is based on the `bluetooth_audio_buttons` plugin.

The following details differ from the `bluetooth_audio_buttons` plugin:
- initialize the listeners from the configuration. I have tried to document the format in the docstring. Is there a better place to do so?
- rpc command is bound with `deference=False`. Otherwise plugins, eg next_song, may not be initialized yet upon initialization. Is there a better way?

s-martin commented:
@pabera @ChisSoc

ChisSoc commented:
Nice addition!

Regarding the documentation: The best way to go would be to write a Ueer Guide Page. The live in `docs/sphinx/userguide`. Just create a new `.rst` file and add to in the `index.rst`. See the bluetooth doc [as an example](https://rpi-jukebox-rfid.readthedocs.io/en/latest/userguide/bluetooth_audio_buttons.html). 

The user guide needs
* How to activate this plugin
* How to find out the device_name that is needed for the configuration
* How to find out the button id
* And a complete configuration example, please

Ah, yes, `dereference`. If I remember correctly this has to do with the order of the loading of the plugins. Try loading your plugin last.

votti commented:
Thanks for the feedback+

Top, I will look into adding the documentation.

@dereference: i already have put the plugin last, but the only way to get
it to run was to add the listeners only when adding them as a "finalize"
step.
Is the performance hit without dereference really bad enough to worry about
this optimization?

Cheers!

On Fri, Dec 30, 2022, 21:06 Chris ***@***.***> wrote:

> Nice addition!
>
> Regarding the documentation: The best way to go would be to write a Ueer
> Guide Page. The live in docs/sphinx/userguide. Just create a new .rst
> file and add to in the index.rst. See the bluetooth doc as an example
> <https://rpi-jukebox-rfid.readthedocs.io/en/latest/userguide/bluetooth_audio_buttons.html>
> .
>
> The user guide needs
>
>    - How to activate this plugin
>    - How to find out the device_name that is needed for the configuration
>    - How to find out the button id
>    - And a complete configuration example, please
>
> Ah, yes, dereference. If I remember correctly this has to do with the
> order of the loading of the plugins. Try loading your plugin last.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/1943#issuecomment-1368074517>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4R2KB5276MQANQHKX3PLTWP46C5ANCNFSM6AAAAAATM5NKVI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ChisSoc commented:
Ah yes, finalize it was. 

Don't worry about the performance impact. It is negligible.

votti commented:
I added now a documentation.
It would be great if someone could test it. As a device eg an USB mouse works well.

pabera reviewed (commented):
I read through your documentation. I found a few typos which I corrected myself rather than calling them out. I committed them to the same branch. I'd like @ChisSoc to do a last check on the Python code.

Great addition. (all though I wonder how my 3 year old would use a USB mouse to rule their Phoniebox 🤔)

votti commented:
Thanks for the comments, I will address them!

`(all though I wonder how my 3 year old would use a USB mouse to rule their Phoniebox thinking)`

The USB mouse was more a very general illustration that this works with any USB device, not necessarily an endorsement :) 
The beter usecase for this is to use a cheap Joystick controller board in order to hook up arcade buttons in via USB in case the GPIO pins are occupied otherwise.
I am using: https://de.aliexpress.com/item/1005001604578101.html?spm=a2g0o.order_list.order_list_main.5.26875c5fq1PFcA&gatewayAdapt=glo2deu

votti commented:
It seems that flake8 is failing due to flake8 violations in code not touched by this PR.
Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

pabera commented:
> Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

A separate PR would be ok.

Would you still be interested in merging this?

AlvinSchiller commented:
> It seems that flake8 is failing due to flake8 violations in code not touched by this PR. Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

For the fixes see https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/1989

votti commented:
Hi @pabera ,

Sorry for being unresponsive, I had quite lot of personal events that diminished my free time.

I rebased the PR to the latest dev branch, so the flake8 should be fixed and the cosmetic issues should be addressed.

pabera commented:
Still seem to have some issues. Have you tried locally?

pabera commented:
I have fixed the flake8 issue in another PR. If you pull all changes again, you should be able to pass flake8 properly

votti commented:
Hi there, 
I rebased again and run `flake8` locally - now all looks clean on my side.

votti commented:
I also found by chance that in the old Jukebox there was already an implementation/reference about this: https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/components/controls/buttons_usb_encoder/README.md?plain=1
Should I link this in the readme?

s-martin commented:
@votti would you mind merging latest changes from `future3/develop`? Biggest change is probably that docs are now markdown.

I think then we could merge this PR

votti commented:
Please have a look at my updates.
I am using this now since a year with a `Zero delay USB encoder` and some arcade buttons and it works well.

s-martin commented on documentation/builders/event-devices.md:
```diff
@@ -0,0 +1,134 @@
+This includes USB peripherals (Keyboards, Controllers, Joysticks or Mouse) as well as potentially bluetooth devices.
+
+A specific usecase for this could be, if a Zero Delay Arcade USB Encoder is used to wire arcade buttons instead of using GPIO pins.
+
```
I would not render the Markdown file as plain.

s-martin replied:
The link has a `?plain=1` at the end. I don't think it's necessary, but it's not that important :)

pabera replied:
I thank he was referring to the addition of `?plain=1` at the end of the URL (your comment on 15-Aug)

votti replied:
Ah sorry now I see it - sorry I was looking at the lines above and could not see anything.
Fixed now. Thanks for pointing this out.

s-martin commented on documentation/builders/event-devices.md:
```diff
@@ -0,0 +1,134 @@
+while watching the logs with `tail -f shared/logs/app.log`.
+Look for entries like `No callback registered for button ...`.
+
+### Specifying the `{rpc_command_defintion}`
```
Typo. An "i" is missing.

s-martin commented on src/jukebox/components/controls/event_devices/README.rst:
```diff
@@ -0,0 +1,131 @@
+Event devices are generic input devices that are exposed in ``/dev/input``.
```
This file could be removed? You put the markdown version under documentation.

pabera commented on documentation/builders/event-devices.md:
```diff
@@ -0,0 +1,134 @@
+
+## Configuration
+
+To configure event devices, first add the plugin as the last entry to the module list:
```
Will it always have to be the last? Why?

pabera commented on documentation/builders/event-devices.md:
````diff
@@ -0,0 +1,134 @@
+
+The `{device_name}` can be identified using the following Python snippet:
+
+``` Python
````
In the spirit of run_*.py files, would it make sense to provide this code as a configurator? Or at least as a runable script so users don't have to write the code themselves?

pabera commented on documentation/builders/event-devices.md:
````diff
@@ -0,0 +1,134 @@
+      event_devices: controls.event_devices # Add as last entry
+```
+
+Then add the following section to the configuration:
````
Similar to [gpio.yaml](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/resources/default-settings/gpio.example.yaml), I recommend to outsource this to it's own file, e.g. `evdev.yaml`. You could then also proovide an `evdev.example.yaml`

pabera commented on documentation/builders/event-devices.md:
```diff
@@ -0,0 +1,134 @@
+
+## Full example config
+
+Here is a complete configuration example for a USB Joystick controller:
```
I wonder, how close can we get this definition to GPIO? It'll be important to maintain parity of functionality between the two input types to reduce confusion with users.

I am referring to 1) the yaml layout (like for [GPIO](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/resources/default-settings/gpio.example.yaml)) and 2) the actions types such es "press", "short/long press", rotary actions etc.

votti replied:
Aiming to harmonize the interface with GPIO seems valuable.

However I doubt that a 1:1 mapping of the config is feasible as the GPIO interface supports a single device with multiple pins, while EVDEV would be multiple devices with multiple pins.

`short/long press` does need additional logic (see discussion here: https://github.com/gvalkov/python-evdev/issues/70).
Things like rotation may be feasible as well, but would need more investigation. In particular it would be helpful to have a device at hand that has such events as a concrete use case.
Output (toggling LEDs, sounds) are also supported by EVDEV, but I dont have a device at hand to test it.

My proposal forward would be that I try to restructure the yaml that it follows the `gpio` interface more closely and maybe add stubs for `short/long press` with an `NotImplementedError`.
Once merged we may add follow up issues to investigate the future directions.

pabera replied:
Sounds like a good idea. If a full harmonization is not possible, I agree it's still valuable to get as close as possible and document the differences.

votti reviewed (commented):
Thanks for the feedback!

To me it is important that we get the interace (`yml` config) right before merging, so I see the following todos for me:

- [x] Factor out the config as `evdev.yml`
- [x] Add `evdev.example.yaml`
- [x] Align the  `evdev.yml` structure with the `gpio.yml`
- [x] Investigate why I indicated that the `evdev` module should be listed last.

The following issues I would like to postpone as follow-up issues after merging:

- [ ] Add configurator script

pabera commented:
@votti We are planning to integrate this PR in our next Release https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2211 which we want to release in the next couple of days. Do you think you will be able to work on the changes suggested or will you need more time? In the second case, we would postpone the release of this PR.

votti commented:
@pabera I am away for holidays so no access to a pi. However I just managed to get the docker dev environment (including evdev devices) runing, so I may be able to work on it till the end of the week.

No hard feelings if this does not make the next release. I think having the configuration file in a final form  before merging it is really important.

votti commented:
I now aligned the configuration structure with `gpio.yaml`.
In contrast to `gpio`, multiple devices each with their own `input/output_devices` are supported.
Note that currently only `input_devices`  the `Button` type with the `on_press` action is implemented.

While extending the config to such new functionalities is now easy, this code is still relying on the pre-existing `EventDeviceListener` code backend. Expanding this mechanisms to support more features (eg long press etc) will require significant changes on this backend. 

" Investigate why I indicated that the evdev module should be listed last." - I did not find a good reason after trying also different orders. Really not sure why I wrote it like this before.

votti commented:
I now also added some tests and fixed some bugs/edge cases found during testing. 

One issue I encountered is that `zmq` is not installed for the Github CI testing.
First I tried to fix the Github CI but then decided to better just use `mocks` to make my tests independent of `zmq`.

I will add an issue that it would be nice to eventually have `zmq` also in our github action workflow used for testing.

pabera commented on resources/default-settings/evdev.example.yaml:
```diff
@@ -0,0 +1,59 @@
+devices: # A list of evdev devices each containing one or multiple input/output devices
+  joystick: # A nickname for a device
+    device_name: DragonRise Inc.   Generic   USB # Device name
```
Is this just a pretty name or actually relevant to access/call the device?

votti replied:
The 'device_name' is relevant as it is used to get the device number (in contrast to the nickname line 2 which is cosmetic).

The docs contain a section how to look it up: https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/49839adb01019cb296d67dc29c06cd9a6af0bc62/documentation/builders/event-devices.md#identifying-the-device_name

pabera commented on test/evdev/test_evdev_init.py:
```diff
@@ -0,0 +1,185 @@
+""" Tests for the evdev __init__ module
```
I like the fact that you have put effort into tests as well. It keeps pushing the bar. It's something we haven't pushed for in the past but indeed allows to be more confident when others make changes in the future.

s-martin commented:
Could be merged?

votti commented:
Let me know if there is anything still required from my side!
If not I would appreciate if this would be merged before it diverges again from the main develop/future3 branch.

pabera reviewed (approved):
Approved. 

Last question @votti .. How do you execute the tests in this scenario?
